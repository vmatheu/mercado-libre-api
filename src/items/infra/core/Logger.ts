import { Injectable, Logger, LoggerService, Scope } from '@nestjs/common';

@Injectable({ scope: Scope.TRANSIENT })
export class LoggerCustom implements LoggerService {
  private prefix: string;
  private logger: Logger;

  init(prefix: string) {
    this.logger = new Logger(prefix);
  }

  log(message: any, ...optionalParams: any[]) {
    this.logger.log(message);
  }
  error(message: any, ...optionalParams: any[]) {
    this.logger.error(message);
  }
  warn(message: any, ...optionalParams: any[]) {
    this.logger.warn(message);
  }
  debug?(message: any, ...optionalParams: any[]) {
    this.logger.debug(message);
  }
  verbose?(message: any, ...optionalParams: any[]) {
    this.logger.verbose(message);
  }
}
