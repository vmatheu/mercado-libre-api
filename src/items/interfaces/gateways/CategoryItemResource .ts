export interface CategoryItemResource {
  getCategories: (categoryId: string) => Promise<string[]>;
}
