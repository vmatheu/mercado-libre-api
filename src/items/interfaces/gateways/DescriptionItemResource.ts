export interface DescriptionItemResource {
  getDescription: (id: string) => Promise<string>;
}
