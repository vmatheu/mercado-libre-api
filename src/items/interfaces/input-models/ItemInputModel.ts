export interface ItemInputModel {
  id: string;
  title: string;
  price: number;
  currencyId: string;
  condition: string;
  picture: string;
  freeShipping: boolean;
  category: string;
}
