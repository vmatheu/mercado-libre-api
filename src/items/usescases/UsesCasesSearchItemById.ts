import { Injectable } from '@nestjs/common';
import { getAuthor } from '../domain/author';
import { buildItemWithSoldAndDesc } from '../domain/Item';
import { LoggerCustom } from '../infra/core/Logger';
import { DescriptionItemResourceEndpoint } from '../infra/endpoints/DescriptionItemResourceEndpoint';
import { ItemResourceEndpoint } from '../infra/endpoints/ItemResourceEndpoint';
import { ItemByIdOutputModel } from '../interfaces/output-models/ItemByIdOutputModel';

@Injectable()
export class UsesCasesSearchItemById {
  constructor(
    private readonly itemResource: ItemResourceEndpoint,
    private readonly descriptionItemResource: DescriptionItemResourceEndpoint,
    private readonly logger: LoggerCustom,
  ) {
    logger.init(UsesCasesSearchItemById.name);
  }

  async getItemById(id: string): Promise<ItemByIdOutputModel> {
    this.logger.debug(`call getItemById=${id}`);
    const item = await this.itemResource.getById(id);
    const description = await this.descriptionItemResource.getDescription(id);

    this.logger.debug(`result getItemById itemI=${JSON.stringify(item)}`);
    return {
      author: getAuthor(),
      item: buildItemWithSoldAndDesc(item, description),
    };
  }
}
