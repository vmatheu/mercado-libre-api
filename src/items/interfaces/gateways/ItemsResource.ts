import { ItemInputModel } from '../input-models/ItemInputModel';

export interface ItemsResource {
  findByQuerySearch: (query: string) => Promise<ItemInputModel[]>;
}
