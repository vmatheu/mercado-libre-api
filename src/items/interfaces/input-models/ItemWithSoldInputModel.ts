import { ItemInputModel } from './ItemInputModel';

export interface ItemWithSoldInputModel extends ItemInputModel {
  soldQuantity: number;
}
