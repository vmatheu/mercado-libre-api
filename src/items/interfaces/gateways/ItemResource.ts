import { ItemWithSoldInputModel } from '../input-models/ItemWithSoldInputModel';

export interface ItemResource {
  getById: (id: string) => Promise<ItemWithSoldInputModel>;
}
