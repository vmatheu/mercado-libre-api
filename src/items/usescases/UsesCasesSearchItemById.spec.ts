import { ItemResource } from '../interfaces/gateways/ItemResource';
import { UsesCasesSearchItemById } from './UsesCasesSearchItemById';
import { mock, mockReset } from 'jest-mock-extended';
import { DescriptionItemResource } from '../interfaces/gateways/DescriptionItemResource';
import { ItemWithSoldInputModel } from '../interfaces/input-models/ItemWithSoldInputModel';
import { LoggerCustom } from '../infra/core/Logger';

describe('UsesCasesSearchItemById', () => {
  const itemResourceMock = mock<ItemResource>();
  const descriptionItemResourceMock = mock<DescriptionItemResource>();
  const loggerMock = mock<LoggerCustom>();

  const itemMock: ItemWithSoldInputModel = {
    id: '123231',
    soldQuantity: 0,
    title: 'Test',
    price: 1009,
    currencyId: 'USD',
    condition: 'new',
    picture: 'no ttiene',
    freeShipping: false,
  };

  beforeEach(async () => {
    mockReset(itemResourceMock);
    mockReset(descriptionItemResourceMock);
  });

  it('should return result item object when call getItemByI', async () => {
    itemResourceMock.getById.mockResolvedValue(itemMock);
    descriptionItemResourceMock.getDescription.mockResolvedValue('description');

    const result = await new UsesCasesSearchItemById(
      itemResourceMock,
      descriptionItemResourceMock,
      loggerMock,
    ).getItemById('id');

    expect(result).toMatchSnapshot();
  });

  it('should return result item with decimals price', async () => {
    const newItemMock: ItemWithSoldInputModel = {
      ...itemMock,
      price: 1000.11,
    };
    itemResourceMock.getById.mockResolvedValue(newItemMock);
    descriptionItemResourceMock.getDescription.mockResolvedValue('description');

    const result = await new UsesCasesSearchItemById(
      itemResourceMock,
      descriptionItemResourceMock,
      loggerMock,
    ).getItemById('id');

    expect(result.item.price.decimals).toEqual(11);
  });
});
