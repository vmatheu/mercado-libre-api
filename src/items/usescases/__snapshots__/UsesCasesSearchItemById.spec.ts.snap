// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsesCasesSearchItemById should return result item object when call getItemByI 1`] = `
Object {
  "author": Object {
    "lastName": "Matheu",
    "name": "Victor",
  },
  "item": Object {
    "condition": "new",
    "description": "description",
    "freeShipping": false,
    "id": "123231",
    "picture": "no ttiene",
    "price": Object {
      "amount": 1009,
      "currency": "USD",
      "decimals": 0,
    },
    "soldQuantity": 0,
    "title": "Test",
  },
}
`;
